Libraries/LibWeb/Editing/Commands.cpp

```diff
@@ -0,0 +1,336 @@
+            continue;
+        }
+
+        // 3. Otherwise, if offset is zero and node is an inline node, or if node is an invisible
+        //    node, set offset to the index of node, then set node to its parent.
+        if ((offset == 0 && is_inline_node(node)) || is_invisible_node(node)) {
+            offset = node->index();
+            node = *node->parent();
+            continue;
+        }
+
+        // 4. Otherwise, if node has a child with index offset − 1 and that child is an editable a,
+        //    remove that child from node, preserving its descendants. Then return true.
+        if (offset_minus_one_child && offset_minus_one_child->is_html_anchor_element() && offset_minus_one_child->is_editable()) {
+            remove_node_preserving_its_descendants(*offset_minus_one_child);
+            return true;
+        }
+
+        // 5. Otherwise, if node has a child with index offset − 1 and that child is not a block
+        //    node or a br or an img, set node to that child, then set offset to the length of node.
+        if (offset_minus_one_child && !is_block_node(*offset_minus_one_child)
+            && !offset_minus_one_child->is_html_br_element() && !offset_minus_one_child->is_html_image_element()) {
+            node = *offset_minus_one_child;
+            offset = node->length();
+            continue;
+        }
+
+        // 6. Otherwise, break from this loop.
+        break;
+    }
+
+    // 5. If node is a Text node and offset is not zero, or if node is a block node that has a child
+    //    with index offset − 1 and that child is a br or hr or img:
+    bool block_node_child_is_relevant_type = false;
+    if (is_block_node(node)) {
+        if (auto* child_node = node->child_at_index(offset - 1)) {
+            auto& child_element = static_cast<DOM::Element&>(*child_node);
+            block_node_child_is_relevant_type = Array { HTML::TagNames::br, HTML::TagNames::hr, HTML::TagNames::img }
+                                                    .contains_slow(child_element.local_name());
+        }
+    }
+    if ((node->is_text() && offset != 0) || block_node_child_is_relevant_type) {
+        // 1. Call collapse(node, offset) on the context object's selection.
+        MUST(selection.collapse(node, offset));
+
+        // 2. Call extend(node, offset − 1) on the context object's selection.
+        MUST(selection.extend(*node, offset - 1));
+
+        // 3. Delete the selection.
+        delete_the_selection(selection);
+
+        // 4. Return true.
+        return true;
+    }
+
+    // 6. If node is an inline node, return true.
+    if (is_inline_node(node))
+        return true;
+
+    // 7. If node is an li or dt or dd and is the first child of its parent, and offset is zero:
+    auto& node_element = static_cast<DOM::Element&>(*node);
+    if (offset == 0 && node->index() == 0
+        && Array { HTML::TagNames::li, HTML::TagNames::dt, HTML::TagNames::dd }.contains_slow(node_element.local_name())) {
```

Comment:
Reading through, but dropping this one now since I see it in quite a few places, but it's a little more typical for us to do:

`child_element.local_name().is_one_of(HTML::TagNames::br, HTML::TagNames::hr, HTML::TagNames::img);`

(like in HTMLParser as an example)

Reply:
Oh good one! Missed that completely. Will fix